src/main/scala/is/hail/io/bgen/BgenWriter.scala

```diff
@@ -0,0 +1,290 @@
+
+      while (dataSize >= 8) {
+        bb += data.toByte
+        data = data >>> 8
+        dataSize -= 8
+        byteIndex += 1
+      }
+      i += 1
+    }
+
+    if (dataSize > 0)
+      bb += data.toByte
+  }
+
+  def sortedIndex(a: Array[Double], idx: Array[Int]) {
+    val n = a.length
+    if (n == 3) {
+      if (a(0) >= a(1) && a(1) >= a(2)) {
+        idx(0) = 0
+        idx(1) = 1
+        idx(2) = 2
+      } else if (a(0) >= a(2) && a(2) >= a(1)) {
+        idx(0) = 0
+        idx(1) = 2
+        idx(2) = 1
+      } else if (a(1) >= a(0) && a(0) >= a(2)) {
+        idx(0) = 1
+        idx(1) = 0
+        idx(2) = 2
+      } else if (a(1) >= a(2) && a(2) >= a(0)) {
+        idx(0) = 2
+        idx(1) = 0
+        idx(2) = 1
+      } else if (a(2) >= a(0) && a(0) >= a(1)) {
+        idx(0) = 1
+        idx(1) = 2
+        idx(2) = 0
+      } else {
+        idx(0) = 2
+        idx(1) = 1
+        idx(2) = 0
+      }
+    } else {
+      throw new UnsupportedOperationException("sortedIndex is only supported for arrays of length 3.")
+    }
+  }
+
+  private def resetIndex(index: Array[Int]) {
+    var i = 0
+    while (i < index.length) {
+      index(i) = i
+      i += 1
+    }
+  }
+
+  def resizeWeights(input: Array[Double], output: ArrayBuilder[Int], resized: Array[Double], index: Array[Int], size: Long) {
+    val n = input.length
+    assert(resized.length == n && index.length == n)
+    resetIndex(index)
+
+    val conversionFactor = size.toDouble / input.sum
+
+    var totalFractional = 0d
+    var i = 0
+    while (i < n) {
+      val x = input(i) * conversionFactor
+      resized(i) = x
+      totalFractional += (x - x.floor)
+      i += 1
+    }
+
+    val F = (totalFractional + 0.5).toInt
+    assert(F >= 0 && F <= n)
+
+    sortedIndex(resized, index)
+
+    i = 0
+    var newSize = 0d
+    while (i < n) {
+      val r = if (index(i) < F) resized(i).ceil else resized(i).floor
+      if (i != n - 1)
+        output += r.toUInt.intRep
+      newSize += r
+      i += 1
+    }
+
+    assert(newSize == size)
+  }
+
+  def resizeWeights(input: Array[Int], output: ArrayBuilder[Int], resized: Array[Double], index: Array[Int], size: Long) {
+    val n = input.length
+    assert(resized.length == n && index.length == n)
+    resetIndex(index)
+
+    val conversionFactor = size.toDouble / input.sum
+
+    var totalFractional = 0d
+    var i = 0
+    while (i < n) {
+      val x = input(i).toUIntFromRep * conversionFactor
+      resized(i) = x
+      totalFractional += (x - x.floor)
+      i += 1
+    }
+
+    val F = (totalFractional + 0.5).toInt
+    assert(F >= 0 && F <= n)
+
+    sortedIndex(resized, index)
+
+    i = 0
+    var newSize = 0d
+    while (i < n) {
+      val r = if (index(i) < F) resized(i).ceil else resized(i).floor
+      if (i != n - 1)
+        output += r.toUInt.intRep
+      newSize += r
+      i += 1
+    }
+
+    assert(newSize == size)
+  }
+
+  def emitHeaderBlock(vds: VariantDataset): Array[Byte] = {
```

Comment:
You could have take the ArrayBuilder, too, but it is only called once (or once per partition once you had `parallel=True`), so I don't really care either way.